core/state_processor_rip7560.go

```diff
@@ -159,51 +177,104 @@ func handleRip7560Transactions(transactions []*types.Transaction, index int, sta
+
+	balanceCheck := new(uint256.Int).Set(preCharge)
 
 	chargeFrom := *st.Sender
-	if st.Paymaster != nil {
+	if st.Paymaster != nil && st.Paymaster.Cmp(common.Address{}) != 0 {
 		chargeFrom = *st.Paymaster
 	}
 
 	if have, want := state.GetBalance(chargeFrom), balanceCheck; have.Cmp(want) < 0 {
-		return &common.Address{}, new(uint256.Int), fmt.Errorf("%w: address %v have %v want %v", ErrInsufficientFunds, chargeFrom.Hex(), have, want)
+		return nil, fmt.Errorf("%w: address %v have %v want %v", ErrInsufficientFunds, chargeFrom.Hex(), have, want)
 	}
 
-	state.SubBalance(chargeFrom, mgval)
-	err := gp.SubGas(mggas.Uint64())
+	state.SubBalance(chargeFrom, preCharge)
+	err := gp.SubGas(preCharge.Uint64())
 	if err != nil {
-		return &common.Address{}, new(uint256.Int), err
+		return nil, err
+	}
+	return preCharge, nil
+}
+
+// refund the transaction payer (either account or paymaster) with the excess gas cost
+func refundPayer(vpr *ValidationPhaseResult, state vm.StateDB, gasUsed uint64, gp *GasPool) {
+	var chargeFrom *common.Address
+	if vpr.PmValidationUsedGas == 0 {
+		chargeFrom = vpr.Tx.Rip7560TransactionData().Sender
+	} else {
+		chargeFrom = vpr.Tx.Rip7560TransactionData().Paymaster
+	}
+
+	actualGasCost := new(uint256.Int).Mul(vpr.EffectiveGasPrice, new(uint256.Int).SetUint64(gasUsed))
+
+	refund := new(uint256.Int).Sub(vpr.PreCharge, actualGasCost)
+
+	state.AddBalance(*chargeFrom, refund)
+
+	gp.AddGas(refund.Uint64())
+}
+
+// CheckNonceRip7560 prechecks nonce of transaction.
+// (standard preCheck function check both nonce and no-code of account)
+func CheckNonceRip7560(tx *types.Rip7560AccountAbstractionTx, st *state.StateDB) error {
+	// Make sure this transaction's nonce is correct.
+	stNonce := st.GetNonce(*tx.Sender)
+	if msgNonce := tx.Nonce; stNonce < msgNonce {
+		return fmt.Errorf("%w: address %v, tx: %d state: %d", ErrNonceTooHigh,
+			tx.Sender.Hex(), msgNonce, stNonce)
+	} else if stNonce > msgNonce {
+		return fmt.Errorf("%w: address %v, tx: %d state: %d", ErrNonceTooLow,
+			tx.Sender.Hex(), msgNonce, stNonce)
+	} else if stNonce+1 < stNonce {
+		return fmt.Errorf("%w: address %v, nonce: %d", ErrNonceMax,
+			tx.Sender.Hex(), stNonce)
 	}
-	return &chargeFrom, mggas, nil
+	return nil
 }
 
-func ApplyRip7560ValidationPhases(chainConfig *params.ChainConfig, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, cfg vm.Config, signingHash common.Hash, estimate ...bool) (*ValidationPhaseResult, error) {
+func ApplyRip7560ValidationPhases(chainConfig *params.ChainConfig, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, cfg vm.Config, estimate ...bool) (*ValidationPhaseResult, error) {
 	var estimateFlag = false
 	if len(estimate) > 0 && estimate[0] {
 		estimateFlag = estimate[0]
 	}
+	err := CheckNonceRip7560(tx.Rip7560TransactionData(), statedb)
```

Comment:
Wouldn't this cause all of the 2d-nonce use cases to fail?

Reply:
Yep I'll fix this in the following PR, thanks!